### PR TITLE
Hw29

### PR DESCRIPTION
**Pull request by frostpv**

### CODE REVIEW

**Review by Victoria-Moiseienko — COMMENTED** [review]
Good job. I have some comments

File: src/main/java/com/internet/shop/controllers/RegistrationController.java [hunk 1]
```diff
@@ -34,6 +37,15 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
...
             user = userService.create(user);
+            Set<Role> roleSet = new HashSet<>();
+            if (user.getId() == 1) {
+                roleSet.add(Role.of("ADMIN"));
+                roleSet.add(Role.of("USER"));
+                user.setRoles(roleSet);
+            } else {
+                roleSet.add(Role.of("USER"));
+                user.setRoles(roleSet);
+            }
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/RegistrationController.java]: let's set only USER role to everyone who registers

and inject Admin in InjectController (with only ADMIN role)

File: src/main/java/com/internet/shop/models/Role.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.internet.shop.models;
+
+public class Role {
+    private Long id;
+    private RoleName roleName;
+
+    public Role(RoleName roleName) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/models/Role.java]:
```suggestion
    private Role(RoleName roleName) {
```

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+public class AuthorizationFilter implements Filter {
+
+    private static final String USER_ID = "user_id";
+    private static final Injector injector = Injector.getInstance("com.internet.shop");
+    private UserService userService = (UserService) injector.getInstance(UserService.class);
+    private Map<String, List<Role.RoleName>> protectedUrls = new HashMap<>();
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        protectedUrls.put("/", List.of(Role.RoleName.USER, Role.RoleName.ADMIN));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: if the url is available to all roles, we can remove it from here

on lines 56-59 you have written the logic to pass the request further in this case

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        protectedUrls.put("/users", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/products/add", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/products", List.of(Role.RoleName.USER, Role.RoleName.ADMIN));
+        protectedUrls.put("/products/edit", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/products/delete", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/cart", List.of(Role.RoleName.USER, Role.RoleName.ADMIN));
+        protectedUrls.put("/cart/add", List.of(Role.RoleName.USER, Role.RoleName.ADMIN));
+        protectedUrls.put("/cart/remove", List.of(Role.RoleName.USER, Role.RoleName.ADMIN));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: this should be available only to user
```suggestion
        protectedUrls.put("/cart", List.of(Role.RoleName.USER));
        protectedUrls.put("/cart/add", List.of(Role.RoleName.USER));
        protectedUrls.put("/cart/remove", List.of(Role.RoleName.USER));
```

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        protectedUrls.put("/user/remove", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/order/add", List.of(Role.RoleName.USER, Role.RoleName.ADMIN));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]:
```suggestion
        protectedUrls.put("/order/add", List.of(Role.RoleName.USER));
```
same

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        protectedUrls.put("/orders", List.of(Role.RoleName.ADMIN));
+        protectedUrls.put("/user/orders", List.of(Role.RoleName.USER, Role.RoleName.ADMIN));
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]:
```suggestion
        protectedUrls.put("/user/orders", List.of(Role.RoleName.USER));
```
same

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest req = (HttpServletRequest) request;
+        HttpServletResponse resp = (HttpServletResponse) response;
+        String servletPath = req.getServletPath();
+        Long userId = (Long) req.getSession().getAttribute(USER_ID);
+
+        if (Objects.isNull(protectedUrls.get(servletPath))) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: let's use `contains()` instead and also put both conditions in one `if`

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            chain.doFilter(req, resp);
+        } else {
+            req.getRequestDispatcher("/WEB-INF/views/errors/denied.jsp")
+                    .forward(req, resp);
+        }
+    }
+
+    @Override
+    public void destroy() {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: redundant empty line

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    private boolean isAuthorized(User user, List<Role.RoleName> authorizedRoles) {
+
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
let's also refactor AuthenticationFilter as proposed in the checklist

File: src/main/java/com/internet/shop/models/Role.java (hunk 2 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/models/Role.java]: fixed

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 3 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: removed

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 4 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: fixed

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 5 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: fixed

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 6 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: fixed

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 8 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: fixed

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 9 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: fixed

File: src/main/java/com/internet/shop/controllers/RegistrationController.java (hunk 1 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/RegistrationController.java]: fixed

File: src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java (hunk 7 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/web/filters/AuthorizationFilter.java]: fixed

File: src/main/java/com/internet/shop/controllers/RegistrationController.java [hunk 10]
```diff
@@ -32,10 +35,14 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp)
         String pswRepeat = req.getParameter("psw-repeat");
         if (psw.equals(pswRepeat)) {
             User user = new User(name, login, psw);
+            Set<Role> roleSet = new HashSet<>();
+            roleSet.add(Role.of("USER"));
+            user.setRoles(roleSet);
             ShoppingCart shoppingCart = new ShoppingCart();
             user = userService.create(user);
-            shoppingCart.setUserId(user.getId());
             shoppingCartService.create(shoppingCart);
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/RegistrationController.java]: let's create a shopping cart when userId is already set

File: src/main/java/com/internet/shop/controllers/InjectController.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        roleSet.add(Role.of("ADMIN"));
+        admin.setRoles(roleSet);
+        User user = new User("Laci", "Laci", "123");
+        roleSet.clear();
+        roleSet.add(Role.of("USER"));
+        user.setRoles(roleSet);
+        ShoppingCart shoppingCart = new ShoppingCart();
+        user = userService.create(user);
+        shoppingCartService.create(shoppingCart);
+        shoppingCart.setUserId(user.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/InjectController.java:58]: same here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/controllers/RegistrationController.java (hunk 10 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/RegistrationController.java]: fixed

File: src/main/java/com/internet/shop/controllers/InjectController.java (hunk 11 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/InjectController.java:58]: fixed

File: src/main/java/com/internet/shop/controllers/InjectController.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        roleSet.add(Role.of("ADMIN"));
+        admin.setRoles(roleSet);
+        User user = new User("Laci", "Laci", "123");
+        roleSet.clear();
+        roleSet.add(Role.of("USER"));
+        user.setRoles(roleSet);
+        user = userService.create(user);
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCartService.create(shoppingCart);
+        shoppingCart.setUserId(user.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/com/internet/shop/controllers/InjectController.java]:
```suggestion
        shoppingCart.setUserId(user.getId());
        shoppingCartService.create(shoppingCart);
```
let's create a shopping cart when **`userId`** is already set

same in the RegistrationController

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/internet/shop/controllers/InjectController.java (hunk 12 above)
Comment (frostpv) [reply src/main/java/com/internet/shop/controllers/InjectController.java]: sory

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**